[WIP] Phylo rpca

kwcantrell opened the pull request:
@cameronmartino  This PR adds support for tax2tree. phylogenetic-rpca (both q2 and standalone) can now accept a taxonomy file and will produce a new taxonomy file that contains the taxonomy for both internal nodes and tips for the tree produced by gemelli. Everything should be in working order however, there are a few outstanding issues that need to be address.

1) tax2tree is now required however I don't believe there is a pip version @wasade ? Should we add installation instructions in the README?

2) The taxonomy file is optional however, I couldn't figure out how to only produce an output taxonomy file in q2 in the case where a taxonomy file was used. Is there a way to have optional outputs in q2? Currently, q2 will produce a "blank" taxonomy file.

3) Is there a taxonomy for the test trees used in the standalone/q2 phylogenetic_rpca tests? I am currently just testing the individual functions. 

Here is an example [empire plot](https://view.qiime2.org/visualization/?type=html&src=https%3A%2F%2Fdl.dropbox.com%2Fs%2Ffw0lklaeuzpbbgj%2Ftest_plot.qzv%3Fdl%3D1) for the 'Moving Pictures' dataset. Internal nodes are still labeled using node numbers however, you can color the biplot using taxonomy.

wasade commented:
I think we should look at getting t2t into pip, i'll make a note to see if i can get it uploaded later this week

gibsramen commented:
@kwcantrell 

Re: (2) there's not a way to specify optional outputs in Q2 yet (see [this](https://forum.qiime2.org/t/specify-when-output-if-optional-vs-required/19541) recent forum discussion). I guess the easiest solution would be to have a separate command for phylo-rpca while generating the new taxonomy file but this definitely isn't ideal.

kwcantrell commented:
Thanks @wasade for getting t2t in pip! I updated the PR to install the pip version of t2t and also split the q2 version of phylo rpca into two different version as suggested by @gibsramen.

cameronmartino commented on gemelli/rpca.py:
```diff
@@ -36,6 +38,66 @@ def phylogenetic_rpca(table: biom.Table,
+    """
+
+    output = phylogenetic_rpca(table=table,
+                              phylogeny=phylogeny,
+                              n_components=n_components,
+                              min_sample_count=min_sample_count,
+                              min_feature_count=min_feature_count,
+                              min_feature_frequency=min_feature_frequency,
+                              min_depth=min_depth,
+                              max_iterations=max_iterations)
+    ord_res, dist_res, phylogeny, counts_by_node, _ = output
+
+    return ord_res, dist_res, phylogeny, counts_by_node
+
+
+def phylogenetic_rpca_with_taxonomy(table: biom.Table,
+                      phylogeny: NewickFormat,
+                      taxonomy: pd.DataFrame,
+                      n_components: Union[int, str] = DEFAULT_COMP,
+                      min_sample_count: int = DEFAULT_MSC,
+                      min_feature_count: int = DEFAULT_MFC,
+                      min_feature_frequency: float = DEFAULT_MFF,
+                      min_depth: int = DEFAULT_MTD,
+                      max_iterations: int = DEFAULT_OPTSPACE_ITERATIONS) -> (
+                          OrdinationResults, DistanceMatrix,
+                          TreeNode, biom.Table, pd.DataFrame):
+    """Runs phylogenetic RPCA.
+
+       This code will be run QIIME 2 versions of gemelli.
+    """
+    taxonomy = taxonomy.to_dataframe()
+    output = phylogenetic_rpca(table=table,
+                              phylogeny=phylogeny,
+                              taxonomy=taxonomy,
+                              n_components=n_components,
+                              min_sample_count=min_sample_count,
+                              min_feature_count=min_feature_count,
+                              min_feature_frequency=min_feature_frequency,
+                              min_depth=min_depth,
+                              max_iterations=max_iterations)
```
@kwcantrell looks like the CI is failing due to the indents in these lines being under-indented for visual indent.

cameronmartino commented:
Looks good, thanks @kwcantrell!
